finality-aleph/src/sync/ticker.rs

```diff
@@ -0,0 +1,115 @@
+use tokio::time::{sleep, Duration, Instant};
+
+/// This struct is used for rate limiting as an on-demand ticker. It can be used for ticking
+/// at least once `max_timeout` but not more than once every `min_timeout`.
+pub struct Ticker {
+    last_tick: Instant,
+    current_timeout: Duration,
+    max_timeout: Duration,
+    min_timeout: Duration,
+}
+
+impl Ticker {
+    pub fn new(max_timeout: Duration, min_timeout: Duration) -> Self {
+        Self {
+            last_tick: Instant::now(),
+            current_timeout: max_timeout,
+            max_timeout,
+            min_timeout,
+        }
+    }
+
+    /// Returns whether at least `min_timeout` time elapsed since last tick.
+    /// If `min_timeout` elapsed since last tick, returns true, sets last tick to now,
+    /// current timout to `max_timeout` and will Return true again if called after `min_timeout`.
+    /// If not, returns false and sets current timeout to `min_timeout`.
```

Comment:
This explanation is a bit too technical, it shouldn't be mentioning current timeouts, but rather what will happen. So I would suggest something like "...returns true and records a tick. If not, calls to `wait` will return when `min_timeout` elapses until the next tick." Similarly the comment in `wait` would be "... When enough time elapsed returns and records a tick."

You are also missing a comment by `new`.